content/src/content/docs/learn/tutorials/hooking-up-a-database.mdx

```diff
@@ -0,0 +1,395 @@
+import { Effect, Layer } from "effect"
+import { createServer } from "node:http"
+import { Client } from "pg";
+const db = new Client({ user: "postgres" });
+//---cut---
+const users = Effect.flatMap(
+  Effect.promise(async () => {
+    const { rows } = await db.query("SELECT * FROM users");
+    return { users: rows };
+  }),
+  HttpServerResponse.json
+);
+```
+
+But it is more idiomatic in Effect to use `pipe`:
+
+```ts twoslash
+import { HttpServerResponse } from "@effect/platform"
+import { Effect, Layer } from "effect"
+import { createServer } from "node:http"
+import { Client } from "pg";
+const db = new Client({ user: "postgres" });
+//---cut---
+const users = Effect.promise(
+  () => db.query("SELECT * FROM users")
+).pipe(
+  Effect.map(({ rows }) => ({ users: rows })),
+  Effect.flatMap(HttpServerResponse.json)
+);
+```
+
+This no longer gives us a type error. We've split our computation into parts:
+fetching the data with `Effect.promise`, manipulating its shape with
+`Effect.map`, and serializing it with `HttpServerResponse.json`. The
+`Effect.flatMap` at the end care of the accidental `Effect` nesting.
+
+## An aside on pipes
+
+If `pipe` isn't quite clicking for you yet, one way to visualise it is that
+you're passing some data through a sequence of functions.
+
+Take this example:
+
+```ts twoslash
+import { pipe } from "effect"
+
+const add = (x: number) => (y: number) => y + x
+const mul = (x: number) => (y: number) => y * x
+const sub = (x: number) => (y: number) => y - x
+
+const result = pipe(
+  1,
+  add(2),
+  mul(3),
+  sub(1)
+)
+console.log(result)
+//=> 8
+```
+
+`add`, `mul`, and `sub` are all functions that return functions. So `add(2)`
+returns a function that can add 2 to a given number: `add(2)(1)` returns 3.
+Our pipeline starts with 1, then passes it through `add(2)`, `mul(3)`, and
+`sub(1)`, resulting in 8.
+
+One of the powers this gives you is the ability to add things easily into
+the pipeline.
+
+```ts twoslash {6-9,14,16,18}
+import { pipe } from "effect"
+
+const add = (x: number) => (y: number) => y + x
+const mul = (x: number) => (y: number) => y * x
+const sub = (x: number) => (y: number) => y - x
+const log = (s: string) => (y: number) => {
+  console.log(s, y)
+  return y
+}
+
+const result = pipe(
+  1,
+  add(2),
+  log("1 + 2 ="),
+  mul(3),
+  log("  * 3 ="),
+  sub(1),
+  log("  - 1 ="),
+)
+```
+
+This prints the following, with `result` still being 8:
+
+```
+1 + 2 = 3
+  * 3 = 9
+  - 1 = 8
+```
+
+As we migrate more complex endpoints, we'll see how this ability to "tap into"
+the pipeline is really useful.
+
+## Adding error handling to our database query
+
+All this talk of error handling might have gotten you thinking: can't our
+database query fail? It can! If the query is malformed, `db.query` will throw an
+error. When that happens, at the moment our endpoint will return a 500 error to
+the client. Change the SQL query to `SELECT * FROM nonexistnet` and try it:
```

Comment:
```suggestion
All this talk of error handling might have gotten you thinking: can't our
database query fail? It can! If the query is malformed, `db.query` will throw an
error. When that happens, our endpoint will currently return a 500 error to
the client. Change the SQL query to `SELECT * FROM nonexistent` and try it:
```